compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,274 @@
+            isCrossinline,
+            isNoinline
+        ).apply {
+            descriptor.bind(this)
+            parent = target
+        }
+    }
+
+    // Create a map for type parameters between source class (Mutable Collection classes) to
+    // target class (class currently in lowering phase), this map is later used for substituting type parameters in generated functions
+    private fun createTypeMap(collectionClass: IrClass, mutableClass: IrClass, targetClass: IrClass): Map<IrTypeParameterSymbol, IrType> {
+        // It is guaranteed by compiler that a class inherits from an immutable collection class must have
+        // some methods overridden, therefore we search for such overriddens, and get substitute type for
+        // our stubs based on these functions. For every collection class we select specific function(s), to make sure all
+        // type parameters are covered by our selection.
+        val targetTypeList = when (collectionClass.name.asString()) {
+            "Iterator", "ListIterator" -> {
+                listOf("next")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.returnType }
+            }
+            "Entry" -> {
+                listOf("<get-key>", "<get-value>")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.returnType }
+            }
+            // Iterable is special, the only way to get type parameter is through iterator() function,
+            // which returns Iterator<T>, therefore we need to look further into type arguments to get the substitution type.
+            "Iterable" -> {
+                listOf("iterator")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .flatMap { function ->
+                        (function.returnType as IrSimpleType).arguments
+                            .map { arg -> (arg as IrTypeProjection).type }
+                    }
+            }
+            "Collection", "List", "Set" -> {
+                listOf("contains")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.valueParameters[0].type }
+            }
+            "Map" -> {
+                listOf("containsKey", "containsValue")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.valueParameters[0].type }
+            }
+            else -> throw IllegalStateException()
+        }
+        if (targetTypeList.isNullOrEmpty() || targetTypeList.size != mutableClass.typeParameters.size) {
+            throw IllegalStateException()
```

Comment:
Please always add description and data from the context to explain what went wrong